---
title: "Antarctica - ASV analysis "
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
---

```{r, echo=FALSE, eval=FALSE}
rmarkdown::render('Antarctica phyloseq.Rmd', output_file = '../docs/Antarctica-phyloseq.html')
```



# Initialize

## Libraries

```{r, echo=TRUE, message=FALSE, warning=FALSE}
   if(any(grepl("package:dvutils", search()))) detach("package:dvutils", unload=TRUE)
   library(dvutils)
   library(stringr)
   library(ggplot2)
   library(dplyr)
   library(tidyr)
   library(tibble)
   library(readr)
   library(maps)
   library(glue)
   library(readxl)
   library(rlang)
   
   library(phyloseq)
   library(vegan)  

   library(knitr)
   library(rmdformats)
   library(kableExtra)

   library(patchwork)

   # library(pr2database)
   # data("pr2")
```

## Markdown
```{r knitr_init, echo=FALSE, cache=FALSE}


## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
              cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               eval = TRUE,
               rows.print=50)
opts_knit$set(width=75)
```


# Get the list of samples for GenBank submission - 2020-07-09 - DO NOT RUN

```{r, eval=FALSE}

genbank <-      metapr2_export_asv(dataset_id_selected = c(16, 17, 18), 
                                   export_phyloseq = FALSE, 
                                   export_long_xls = FALSE,
                                   export_wide_xls = FALSE,
                                   export_fasta = FALSE,
                                   export_sample_xls = TRUE,
                                   boot_min = 90, 
                                   boot_level = supergroup_boot,
                                   directory = "../genbank/")

```


# Get the Antarctica ASVs


|set id |  description |
|---|---|
| 16 | Antar_2015_18S_V4 |
| 17 | Antar_2015_16S_plastid |
| 18 | Antar_2015_18S_V4_sorted |
 
* Only use asv for which supergroup_boot >= 90
* Only keep photosynthetic groups abd exclude dinoflagellates


## Initialize constants

```{r}
  # rm(list = ls())
  sample_type <- c("18S filter", "16S plastid", "18S sort")
  taxon_selected <- c("Chlorophyta", "Cryptophyta", "Rhodophyta","Haptophyta", "Ochrophyta")
  class_excluded <- c("Chrysophyceae", "Bangiophyceae", "Florideophyceae", "Xanthophyceae", "Phaeophyceae")

  ps <- list()
  long <- list()
```


## Read the colors

```{r define_colors}

 class_colors <- read_excel("colors.xlsx", sheet = "classes") %>% 
  filter(!(class %in% class_excluded))
  
 class_colors <-structure(class_colors$color_name,.Names=class_colors$class) 
 
 genus_colors <- read_excel("colors.xlsx", sheet = "genus")%>% 
  filter(!(class %in% class_excluded))
  
 genus_colors <-structure(genus_colors$color_name,.Names=genus_colors$genus)  
 
```


## Read the data from the database (only done once) - DO NOT RUN !

* After running rename "_Chlor_Crypt_Rhodo_Hapto_Ochro" to "_photo"

```{r, eval=FALSE}
  
# Export as specific data set as a phyloseq file

  for (i in 1:3) {
      
      temp <- metapr2_export_asv(dataset_id_selected = i + 15,
                                                   taxo_level = division,
                                                   taxo_name = taxon_selected,
                                                   export_phyloseq = TRUE,
                                                   export_long_xls = FALSE,
                                                   export_wide_xls = TRUE,
                                                   export_fasta = TRUE,
                                                   export_sample_xls = TRUE,
                                                   boot_min = 90,
                                                   boot_level = supergroup_boot,
                                                   directory = "../dada2/")


  }


  rm(temp)


```

## Read the data from the database (only done once) using the hash value as asv name - DO NOT RUN !

```{r, eval=FALSE}
  
# Export as specific data set as a phyloseq file

    temp <- metapr2_export_asv(dataset_id_selected = c(16,18),
                                                   taxo_level = division,
                                                   taxo_name = taxon_selected,
                                                   export_phyloseq = FALSE,
                                                   export_long_xls = FALSE,
                                                   export_wide_xls = TRUE,
                                                   export_fasta = FALSE,
                                                   export_sample_xls = FALSE,
                                                   boot_min = 90,
                                                   boot_level = supergroup_boot,
                                                   use_hash = TRUE,
                                                   directory = "../dada2/hash/")



  rm(temp)


```


## Read the Phyloseq files from the disk
* Only keep station 6
* Do not consider TFF samples 

```{r, message=FALSE, eval = TRUE}

  for (i in 1:3) {
    
      ps[[ sample_type[i] ]] <- readRDS(str_c("../dada2/phyloseq_metapr2_asv_set_", i+15, "_photo.rda") )
      
      
  # Filter out station 14
      ps[[ sample_type[i] ]] <- subset_samples(ps[[ sample_type[i] ]], station_id=="6")
      
  # Filter out TFF concentrated samples
      if (i==3) {
      ps[[ sample_type[i] ]] <- subset_samples(ps[[ sample_type[i] ]], is.na(sample_concentration))  
      }
      
      
       ps[[ sample_type[i] ]] <-  ps[[ sample_type[i] ]] %>% 
         
  # Filter ASVs that are absent
        filter_taxa(function(x) sum(x) > 0 , TRUE) %>% 
         
  # Filter Calls to be removed
        subset_taxa(!(class %in% class_excluded)) 

  # Label samples with date 
      sample_data(ps[[ sample_type[i] ]])$sample_label <- as.character(sample_data(ps[[ sample_type[i] ]])$date)
      
  # Add TFF to sample name
      # sample_data(ps[[one_sample_type]])$sample_label <- str_c(sample_data(ps[[one_sample_type]])$metadata_code, 
      #                                                          str_replace_na(sample_data(ps[[one_sample_type]])$sample_concentration, ""), 
      #                                                          sep = "." )
      
      print(glue("Phyloseq - {sample_type[i]}"))
      print(ps[[ sample_type[i] ]] )
      print( glue("Mean number of reads: {mean(sample_sums(ps[[ sample_type[i] ]]))}") )
      cat("============================\n")
      mean(sample_sums(ps[[ sample_type[i] ]]))  
      
  }
  
  
  sample_variables(ps[["18S filter"]])
  mean(sample_sums(ps[["18S filter"]]))  

```



# Get the different fraction separately

```{r, message=FALSE}


  fraction_filter <- c(0.2, 3, 20)
  fraction_sort <- c("pico", "nano")
  
  sample_type_filtered <- c(str_c(sample_type[1], fraction_filter, "um", sep=" "), 
                            str_c(sample_type[2], fraction_filter, "um", sep=" "),
                            str_c(sample_type[3], fraction_sort, sep=" "))
  
# Export as specific data set as a phyloseq file
  
  for (one_sample_type in sample_type[1:2]) {
    for (one_fraction in fraction_filter) {
      one_sample_type_filtered <- c(str_c(one_sample_type, one_fraction, "um", sep=" "))
      ps[[one_sample_type_filtered]] <- ps[[one_sample_type]] %>% 
        subset_samples(fraction_min == one_fraction) %>% 
        filter_taxa(function(x) sum(x) > 0 , TRUE)
      
      print(glue("Phyloseq - {one_sample_type_filtered}"))
      print(ps[[one_sample_type_filtered]] )
      cat("============================\n")
    }
  
  }
  
# Sorted samples
  
    for (one_sample_type in sample_type[3]) {
    for (one_fraction in fraction_sort) {
      one_sample_type_filtered <- c(str_c(one_sample_type, one_fraction, sep=" "))
      ps[[one_sample_type_filtered]] <- ps[[one_sample_type]] %>% 
        subset_samples(fraction_name == one_fraction) %>% 
        filter_taxa(function(x) sum(x) > 0 , TRUE)
      
      print(glue("Phyloseq - {one_sample_type_filtered}"))
      print(ps[[one_sample_type_filtered]] )
      cat("============================\n")
    }
  
    }
  
  sample_type_all <- c(sample_type, sample_type_filtered)
  

```
# Surface samples

* Only surface considered (5 m)
* Very important, must remove taxa that are not present in the filtered samples

```{r}
  sample_type_surface <- str_c(sample_type_all, " surface")
  
  for (one_sample_type in sample_type_all) {
      
      one_sample_type_surface = str_c(one_sample_type, " surface")

      ps[[one_sample_type_surface]] <- ps[[one_sample_type]] %>% 
        subset_samples(depth_level=="surface")  %>%
        filter_taxa(function(x) sum(x) > 0 , TRUE)
      
      # Rename the samples
      
      if (!str_detect(one_sample_type_surface, "sort")){
        sample_names(ps[[one_sample_type_surface]]) <- str_c(as.character(sample_data(ps[[one_sample_type_surface]])$date), 
                                               as.character(sample_data(ps[[one_sample_type_surface]])$depth), "m",
                                               as.character(sample_data(ps[[one_sample_type_surface]])$fraction_min),
                                               sep="_")
      } else {
        sample_names(ps[[one_sample_type_surface]]) <- str_c(as.character(sample_data(ps[[one_sample_type_surface]])$date), 
                                       as.character(sample_data(ps[[one_sample_type_surface]])$depth), "m",
                                       as.character(sample_data(ps[[one_sample_type_surface]])$fraction_name),
                                       sep="_")
      }
      
      
      print(glue("Phyloseq - {one_sample_type_surface}"))
      print(ps[[one_sample_type_surface]] )
      cat("============================\n")
      
  }

```
## Normalize and transform to long form

```{r, message=FALSE}

  for (one_sample_type in sample_type_surface) {
    
      cat(one_sample_type)
      cat("\n============================")  
      
      ps[[one_sample_type]] <- phyloseq_normalize_median(ps[[one_sample_type]])
      
      long[[one_sample_type]] <- phyloseq_transform_to_long(ps[[one_sample_type]])
      
      cat("\n") 
      
  }


```

## List of classes

```{r fig.height=6, fig.width=6}

  classes <- data.frame(division = NULL, class= NULL)
  
  for (one_sample_type in sample_type_surface[1:3]) {
      classes <- bind_rows(classes, select(long[[one_sample_type]], division, class))
  }

      classes <- classes %>% 
        distinct(division, class) %>% 
        arrange(division, class)
      
      # print(classes, n=50)
      knitr::kable(classes)

```


## Treemaps


```{r fig.height=6, fig.width=6}


  treemap_class <- list()
  treemap_genus <- list()
  
  for (one_sample_type in sample_type_surface) {
    
      treemap_class[[one_sample_type]] <- phyloseq_long_treemap(long[[one_sample_type]],
                                                                     division, class,
                                                                     str_c(one_sample_type, " - Class" ),
                                                                     colors = class_colors,
                                                                     label_group1 = FALSE) 

      treemap_genus[[one_sample_type]] <- phyloseq_long_treemap(long[[one_sample_type]],
                                                                     class, genus,
                                                                     str_c(one_sample_type, " - Genus" ),
                                                                     colors = genus_colors)

  }

```

## ASVs bargraphs


```{r, fig.height=8, fig.width=8}

  bargraph_asv <- list()
  bargraph_species <- list()
  
  for (one_sample_type in sample_type_surface) {
    
      bargraph_asv [[one_sample_type]] <- phyloseq_long_bargraph(long[[one_sample_type]], text_scaling = 0.75,n_bars = 20, use_asv = TRUE, 
                                                                title = one_sample_type,
                                                                taxo_level_fill = class, taxo_colors_fill = class_colors)
      bargraph_species[[one_sample_type]] <- phyloseq_long_bargraph(long[[one_sample_type]], text_scaling = 0.75,n_bars = 20 ,use_asv = FALSE, 
                                                                title = one_sample_type,
                                                                taxo_level_fill = class, taxo_colors_fill = class_colors)

  }


```
## Barplots per sample

```{r fig.height=8, fig.width=8}

  bargraph_sample <- list()

  for (one_sample_type in sample_type_surface[4:11]) { 
    
    bargraph_sample[[one_sample_type]] <- plot_bar(ps[[one_sample_type]], x = "sample_label", fill = "class") + 
        geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") + 
        ggtitle(str_c("Class level - ", one_sample_type)) + theme(axis.text.y = element_text(size = 10)) + 
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) + coord_flip() + 
        scale_fill_viridis_d() + scale_color_viridis_d()
    print(bargraph_sample[[one_sample_type]])
}
```

## Heatmaps

Filter : Dictyo / Chrysophyceae /Pelago /Bacili / Crypto / Pyrami / Mamiello

### Class

#### Most abundant 10%

```{r fig.height=6, fig.width=9}
  heatmap_class_abundant <- list()

  for (one_sample_type in sample_type_surface[4:11]) { 
    
    ps_heat <- tax_glom(ps[[one_sample_type]],taxrank = "class") %>% 
      phyloseq_filter_abundant_taxa(fraction_min=0.1)
      
    heatmap_class_abundant[[one_sample_type]]  <- plot_heatmap(ps_heat, method = "NMDS", distance = "bray", 
                 taxa.label = "class",  taxa.order="division",
                 sample.label = "sample_label", sample.order="sample_label",
                 low="beige", high="red", na.value="beige", trans = NULL,
                 title=one_sample_type) + 
                 xlab("") + ylab("")
    
    # plot(heatmap(otu_table(ps_heat)))
    
    print(heatmap_class_abundant[[one_sample_type]] )
  
  }

```

#### Selected
```{r fig.height=6, fig.width=9}

class_selected <- c("Dictyochophyceae", "Pelagophyceae", "Bacillariophyta", "Prymnesiophyceae", "Cryptophyceae", "Pyramimonadophyceae", "Mamiellophyceae")

  heatmap_class_selected <- list()

  for (one_sample_type in sample_type_surface[4:11]) {
  # for (one_sample_type in sample_type_surface[4:4]) {
    reads_max <- case_when(str_detect(one_sample_type, "18S filter") ~ 10000,
                           str_detect(one_sample_type, "16S plastid") ~ 30000,
                           str_detect(one_sample_type, "18S sort") ~ 10000)
    
    if (str_detect(one_sample_type, "18S sort")) {
        class_selected <- c("Pelagophyceae", "Bacillariophyta", "Prymnesiophyceae", "Cryptophyceae", "Mamiellophyceae") 
        }
    
    ps_heat <- tax_glom(ps[[one_sample_type]],taxrank = "class") %>% 
      subset_taxa(class %in% class_selected )
    
    # Try to order by division and class...
    # tax_table <- data.frame(tax_table(ps_heat)@.Data)
    # taxa_names(ps_heat) <- str_c(tax_table$division, tax_table$class, sep=" - ") 
      
   gg  <- plot_heatmap(ps_heat, method = "NMDS", distance = "bray", 
                       taxa.label = "class",  taxa.order="class",
                       sample.label = "sample_label", sample.order="sample_label",
                       low="beige", high="red", na.value="red", trans=NULL, 
                       title=one_sample_type) +
         theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust=1),
               axis.title.x=element_blank(), 
               axis.title.y=element_blank(),
               panel.background = element_rect(fill = "grey50"),
               panel.grid.major = element_blank()) +
         scale_fill_gradient(limits= c(0, reads_max), low="beige", high="red", na.value="red") 
    
    if (str_detect(one_sample_type,"18S filter 0.2")) gg <- gg +
     expand_limits(y= c(0.5,8) )     +
     geom_vline(xintercept = c(3.5, 9.5, 12.5)) + 
     annotate(geom = "text", label = c("summer", "fall", "spring", "summer"), x= c(2, 6.5, 11, 15), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))  
     

    if (one_sample_type == "18S filter 3 um surface") gg <- gg +
     expand_limits(y= c(0.5,8) )    +  
     geom_vline(xintercept = c(4.5, 10.5, 13.5))  +
     annotate(geom = "text", label = c("summer", "fall", "spring", "summer"), x= c(2.5, 7.5, 12, 16), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))

    if (one_sample_type == "18S filter 20 um surface") gg <- gg +
     expand_limits(y= c(0.5,8) )    +  
     geom_vline(xintercept = c(4.5, 9.5)) +
     annotate(geom = "text", label = c("summer", "fall", "summer"), x= c(2.5, 7, 12.5), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))
   
    if (str_detect(one_sample_type,"16S plastid 0.2")) gg <- gg +
     expand_limits(y= c(0.5,8) )     +
     geom_vline(xintercept = c(4.5, 5.5, 7.5)) + 
     annotate(geom = "text", label = c("summer", "fall", "spring", "summer"), x= c(2.5, 5, 6.5, 10), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))  
     

    if (one_sample_type == "16S plastid 3 um surface") gg <- gg +
     expand_limits(y= c(0.5,8) )    +  
     geom_vline(xintercept = c(3.5, 8.5, 11.5))  +
     annotate(geom = "text", label = c("summer", "fall", "spring", "summer"), x= c(2, 6, 10, 14), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))

    if (one_sample_type == "16S plastid 20 um surface") gg <- gg +
     expand_limits(y= c(0.5,8) )    +  
     geom_vline(xintercept = c(4.5, 8.5, 9.5))  +
     annotate(geom = "text", label = c("summer", "fall", "spring", "summer"), x= c(2.5, 6, 9, 11.5), y = 7.65, 
              color = "white", fill= "grey50", 
              hjust = 0.5, vjust = 0, size = 4.5, label.r = unit(0.20, "lines"))

    
    # plot(heatmap(otu_table(ps_heat)))
    # , trans = scales::log_trans(10)
    
    print(gg)
    
    heatmap_class_selected[[one_sample_type]] <- gg
  
  }
```


### Species

#### Most abundant

```{r fig.height=6, fig.width=9}
  heatmap_genus_abundant <- list()

  for (one_sample_type in sample_type_surface[4:11]) { 
    
    ps_heat <- tax_glom(ps[[one_sample_type]],taxrank = "species") %>% 
      phyloseq_filter_abundant_taxa(fraction_min=0.1)
    
    reads_max <- case_when(str_detect(one_sample_type, "18S filter") ~ 10000,
                           str_detect(one_sample_type, "16S plastid") ~ 5000,
                           str_detect(one_sample_type, "18S sort") ~ 10000)    

      
    heatmap_genus_abundant[[one_sample_type]]  <- plot_heatmap(ps_heat, method = "NMDS", distance = "bray", 
                 taxa.label = "species",  taxa.order="division",
                 sample.label = "sample_label", sample.order="sample_label",
                 low="beige", high="red", na.value="red", 
                 title=one_sample_type)  +
                 scale_fill_gradient(limits= c(0,reads_max), low="beige", high="red", na.value="red")
    
    print(heatmap_genus_abundant[[one_sample_type]] )
  
  }
```

#### Selected

```{r fig.height=6, fig.width=9}

species_selected <- c("Thalassiosira_minima","Thalassiosira_antarctica", "Fragilariopsis_cylindrus",  "Minidiscus_sp.", "Chaetoceros_neogracilis",
                      "Porosira_glacialis", "Coretrhon_inerme", "Palmaria_palmata", "Pseudo-nitzchia_seriata",
                      "Micromonas_polaris",  "Micromonas_clade_B3","Bathycoccus_prasinos", "Pyramimonas_gelidicola", 
                      "Geminigera_cryophila", "Pelagophyceae_XXX_sp." ,
                      "Phaeocystis_antarctica")


  heatmap_species_selected <- list()

  for (one_sample_type in sample_type_surface[4:11]) {
  # for (one_sample_type in sample_type_surface[4:4]) { 
    reads_max <- case_when(str_detect(one_sample_type, "18S filter") ~ 10000,
                           str_detect(one_sample_type, "16S plastid") ~ 5000,
                           str_detect(one_sample_type, "18S sort") ~ 10000)    
    if (str_detect(one_sample_type, "18S sort")) {
        species_selected <- c("Thalassiosira_minima", "Fragilariopsis_cylindrus",  "Minidiscus_sp.", "Chaetoceros_neogracilis",
                      "Porosira_glacialis", "Coretrhon_inerme", "Palmaria_palmata", "Pseudo-nitzchia_seriata",
                      "Micromonas_polaris",   
                      "Geminigera_cryophila", "Pelagophyceae_XXX_sp." ,
                      "Phaeocystis_antarctica") 
        }

    if (str_detect(one_sample_type, "16S plastid")) {
        species_selected <- c("Minidiscus_sp.", "Porosira_glacialis", "Pseudo-nitzchia_sp.", "Thalassiosira_sp.",
                              "Phaeocystis_sp.", "Chrysochromulina_sp.",
                              "Pelagomonadaceae_X_sp.", "Cryptomonadales_XX_sp.",
                              "Pyramimonas_sp.") 
        }

    ps_heat <- tax_glom(ps[[one_sample_type]],taxrank = "species") %>% 
      subset_taxa(species %in% species_selected )
    
    # Try to order by division and species...
    # tax_table <- data.frame(tax_table(ps_heat)@.Data)
    # taxa_names(ps_heat) <- str_c(tax_table$division, tax_table$species, sep=" - ") 
      
    gg  <- plot_heatmap(ps_heat, method = "NMDS", distance = "bray", 
                       taxa.label = "species",  taxa.order="class",
                       sample.label = "sample_label", sample.order="sample_label",
                       low="beige", high="red", na.value="red", trans = scales::log10_trans(),
                       title=one_sample_type) +
                       theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust=1),
                             axis.title.x=element_blank(), axis.title.y=element_blank(),
                             panel.background = element_rect(fill = "grey50"),
                             panel.grid.major = element_blank())  +
                       scale_fill_gradient(limits= c(0,reads_max), low="beige", high="red", na.value="red")
      
    if (str_detect(one_sample_type, "18S filter 0.2")) gg <- gg + geom_vline(xintercept = c(3.5, 9.5, 12.5))

    if (one_sample_type == "18S filter 3 um surface") gg <- gg +  geom_vline(xintercept = c(4.5, 10.5, 13.5))

    if (one_sample_type == "18S filter 20 um surface") gg <- gg +  geom_vline(xintercept = c(4.5, 9.5))
    
    if (str_detect(one_sample_type,"16S plastid 0.2")) gg <- gg +  geom_vline(xintercept = c(4.5, 5.5, 7.5))
     

    if (one_sample_type == "16S plastid 3 um surface") gg <- gg + geom_vline(xintercept = c(3.5, 8.5, 11.5)) 
    
    if (one_sample_type == "16S plastid 20 um surface") gg <- gg + geom_vline(xintercept = c(4.5, 8.5, 9.5))
                
    
    # plot(heatmap(otu_table(ps_heat)))
    # , trans = scales::log_trans(10)
    
    print(gg )
    
    heatmap_species_selected[[one_sample_type]] <- gg
  
  }
```

# NMDS


## Apply to all samples

```{r}
  nmds_sample <- list()

  for (one_sample_type in sample_type_surface[1:3]) { 
    
    one_ps <- ps[[one_sample_type]] %>% 
      tax_glom(taxrank = "species") 
    
    nmds_sample[[one_sample_type]] <- phyloseq_nmds(one_ps, 
                                                    title=one_sample_type, 
                                                    sample_color = season, 
                                                    sample_shape = fraction_name_original,
                                                    taxo_level = class,
                                                    taxo_colors = class_colors)
  }



```

# ANOSIM and PERMANOVA

https://joey711.github.io/phyloseq-demo/Restroom-Biogeography.html

## Apply to all samples

```{r}

  for (one_sample_type in sample_type_surface[c(1,2)]) { 
    
   cat( "\n", "===========================", "\n")
   cat (glue::glue("Data set: {one_sample_type}"), "\n")
   cat("===========================", "\n")
    
    one_ps <- ps[[one_sample_type]] %>% 
      tax_glom(taxrank = "species") 
    
# ANOSIM
    for (one_factor in c("season", "fraction_name_original")) {
    
    group = phyloseq::get_variable(one_ps, one_factor)
    
    anosim = vegan::anosim(distance(one_ps, "bray"), group)
    
    cat (glue::glue("ANOSIM - Variable: {one_factor}"), "\n")
    cat (glue::glue("{unique(group)} "), "\n")
    

    print(anosim)
    }
    
# Permanova
    
  bray.dist <- phyloseq::distance(one_ps, method = "bray")

  # making a data frame from the sample_data

  sample.df <- data.frame(sample_data(one_ps))  
  
  cat (glue::glue("PERMANOVA"), "\n")
    
  print(vegan::adonis(bray.dist ~ season + fraction_name_original, data = sample.df))

  }



```



# Vertical profile 

* Station 6 - 2015-01-16 
* Very important, must remove taxa that are not present in the filtered samples


## Filter the data
```{r}
  sample_type_profile <- str_c(sample_type_all, " profile")
  
  for (one_sample_type in sample_type_all) {
      
      one_sample_type_profile = str_c(one_sample_type, " profile")

      ps[[one_sample_type_profile]] <- ps[[one_sample_type]] %>% 
        subset_samples(station_id=="6") %>% 
        subset_samples(date =="2015-01-16") %>%
        filter_taxa(function(x) sum(x) > 0 , TRUE) %>% 
        phyloseq_normalize_median()
      
      # Rename the samples
      
      if (!str_detect(one_sample_type_profile, "sort")){
        sample_names(ps[[one_sample_type_profile]]) <- str_c(as.character(sample_data(ps[[one_sample_type_profile]])$depth), "m",
                                               as.character(sample_data(ps[[one_sample_type_profile]])$fraction_min),
                                               sep="_")
      } else {
        sample_names(ps[[one_sample_type_profile]]) <- str_c(as.character(sample_data(ps[[one_sample_type_profile]])$depth), "m",
                                       as.character(sample_data(ps[[one_sample_type_profile]])$fraction_name),
                                       sep="_")
      }
      
      
      print(glue("Phyloseq - {one_sample_type_profile}"))
      print(ps[[one_sample_type_profile]] )
      cat("============================\n")
      
  }

```

## Barplots per depth

```{r fig.height=8, fig.width=8}

  bargraph_sample <- list()

  for (one_sample_type in sample_type_profile[4:11]) { 
    
    bargraph_sample[[one_sample_type]] <- plot_bar(ps[[one_sample_type]], x = "depth", fill = "class") + 
        geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") + 
        ggtitle(str_c("Class level - ", one_sample_type)) + theme(axis.text.y = element_text(size = 10)) + 
        theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) + coord_flip() + 
        # scale_fill_viridis_d() + scale_color_viridis_d() + 
        scale_fill_manual(values = class_colors) +  scale_color_manual(values = class_colors) + 
        scale_x_reverse() + xlab("Depth (m)") + ylab("Reads") +
        theme_bw()
    
    print(bargraph_sample[[one_sample_type]])
}
```

# Compare the contributions of CLASS, GENUS and SPECIES using the different methods

## Sample by sample comparison

* Each dot corresponds to the relative contribution of the taxonomic level considered for a sample for which both 18S and 16S has been performed

```{r fig.height=6, fig.width=10}
long_species_sample <- list()
long_class_sample <- list()

  for (one_sample_type in sample_type) {
      ps_one <- phyloseq_normalize_percent(ps[[one_sample_type]])
      
      long_one <- phyloseq_transform_to_long(ps_one) %>%
        filter (!is.na(fraction_name_original)) %>% 
        group_by(date, station_id, depth, fraction_name_original, division, class, order, family, genus, species) %>%
        summarize(n_reads = sum(n_reads)) %>% 
        ungroup()
      
      long_species_sample[[one_sample_type]] <- long_one 
      
      long_one <- phyloseq_transform_to_long(ps_one) %>%
        filter (!is.na(fraction_name_original)) %>%
        group_by(date, station_id, depth, fraction_name_original, division, class) %>%
        summarize(n_reads = sum(n_reads)) %>% 
        ungroup()
      
      long_class_sample[[one_sample_type]] <- long_one 
    }
   
 long_compare_species_18S_16S <- full_join(rename(long_species_sample[["18S filter"]], n_reads_18S_filter = n_reads),
                                           rename(long_species_sample[["16S plastid"]],n_reads_16S_plastid = n_reads))
 
 ggplot(long_compare_species_18S_16S, aes(x = n_reads_18S_filter, y = n_reads_16S_plastid)) +
   geom_point(aes(color=class)) +
   scale_color_manual(values = class_colors) +
   ggtitle("Species")
 
 long_compare_class_18S_16S <- long_compare_species_18S_16S %>% 
        group_by(date, station_id, depth, fraction_name_original, division, class) %>%
        summarize(n_reads_18S_filter = sum(n_reads_18S_filter, na.rm=TRUE), n_reads_16S_plastid = sum( n_reads_16S_plastid, na.rm=TRUE)) 
 
  ggplot(long_compare_class_18S_16S, aes(x = n_reads_18S_filter, y = n_reads_16S_plastid)) +
   geom_point(aes(color=class)) +
   scale_color_manual(values = class_colors)+
   ggtitle("Classes")
   
```
## Build the species table including all samples

**TO DO**: Remove XXX sp. from euphotic zone if another species from that genus in the surface table (e.g. Micromonas sp.)

```{r}
unfill <- function(x) {
  same <- x == dplyr::lag(x)
  ifelse(!is.na(same) & same, NA, x)
}
```


```{r}
  table_species <- list()

  for (one_sample_type in sample_type) {
    
      table_species [[one_sample_type]] <- long_species_sample[[one_sample_type]] %>% 
        mutate(level= ifelse(depth == 5, "surface", "euphotic")) %>% 
        group_by(division, class, order, family, genus, species, level) %>%
        summarize(min_pct_reads = min(n_reads, na.rm = TRUE)*100, 
                  mean_pct_reads = mean(n_reads, na.rm = TRUE)*100, 
                  max_pct_reads = max(n_reads, na.rm = TRUE)*100, 
                  n_samples = n()) %>% 
        rename_at(vars(matches("n_|pct_")), funs(str_c( .,str_replace_all(one_sample_type, " ", "_"), sep="_"))) %>% 
        ungroup() 
      
  }

  table_species <- table_species %>% 
    purrr::reduce(full_join) %>% 
    select(-order, -family, -genus) %>% 
    mutate(species = str_replace(species, "_X+", "")) %>% 
    arrange(division, class, species) %>% 
    filter(!str_detect(species, "ales|phyceae|phyta|aceae|Raphid-|MOCH|Luka"))
  
  table_species_surface <- table_species %>% 
    filter (level == "surface") %>%
    select(-level) %>% 
    mutate(across(c("division", "class"), ~ unfill(.x))) %>% 
    mutate(across(c("division", "class", "species"), ~ str_replace_all(.x, "_", " ")))

  table_species_euphotic_only <- table_species %>% 
    select(division, class, species, level) %>% 
    mutate(rn = row_number()) %>% 
    pivot_wider(names_from = level, values_from = rn ) %>% 
    filter(is.na(surface)) %>% 
    select(-surface, -euphotic) %>% 
    mutate(across(c("division", "class"), ~ unfill(.x))) %>% 
    mutate(across(c("division", "class", "species"), ~ str_replace_all(.x, "_", " ")))
    
  
  
  openxlsx::write.xlsx(table_species_surface, 
                        file="../dada2/species_table_surface.xlsx",
                        sheetName=c("Species"),
                        firstActiveRow = 2, colWidths ="auto", numFmt="0.00", zoom=75)
  
  openxlsx::write.xlsx(table_species_euphotic_only, 
                        file="../dada2/species_table_euphotic.xlsx",
                        sheetName=c("Species"),
                        firstActiveRow = 2, colWidths ="auto", numFmt="0.00", zoom=75)  
  
  

```



## Define upset function
* Needs only to have 0 and 1 in the matrix

```{r}
plot_upset_long <- function(long_comparison, file_name){

long_upset <- long_comparison %>% 
  mutate_if(is.numeric, funs(case_when(is.na(.) ~ 0,
                                       TRUE ~ 1))) %>% 
  as.data.frame()

fig_upset <- UpSetR::upset(long_upset, empty.intersections = "on", number.angles = 30, point.size = 12, line.size = 4, 
    mainbar.y.label = "Taxa number", sets.x.label = "Taxa #", order.by = "freq",
    text.scale = c(5), mb.ratio = c(0.7, 0.3))

# text.scale = c(intersection size title, intersection size tick labels, set size title, set size tick labels, set names, numbers above bars)

print(fig_upset)
png(file_name, width = 2000, height = 1200, units = "px")
print(fig_upset)
dev.off()

return(fig_upset)
}
```


## Global comparison of species found with the 3 methods


* Compute table of number of samples for each species (rows) vs the three methods (columns)

* Only keep
  * 2015 samples because it is the only dataset for which we have the three types of samples
  * 0.2 and 3 um fractions (to be comparable with sorting)
* Remove species that contains _X


```{r  fig.height=12, fig.width=20}

long_species <- list()
long_genus <- list()
long_class   <- list()
fig_upset_list <- list()

  for (one_sample_type in sample_type) {
    
      long_species_sample_filtered <- long_species_sample[[one_sample_type]] %>%
        filter(date >= as.Date("2015-01-01")) %>% 
        filter(fraction_name_original != "> 20 um")
    
      long_one <- long_species_sample_filtered %>% 
        group_by(division, class, order, family, genus, species) %>%
        summarize(min_pct_reads = min(n_reads, na.rm = TRUE)*100, 
                  mean_pct_reads = mean(n_reads, na.rm = TRUE)*100, 
                  max_pct_reads = max(n_reads, na.rm = TRUE)*100, 
                  n_samples = n()) %>% 
        rename_at(vars(matches("n_|pct_")), funs(str_c( .,str_replace_all(one_sample_type, " ", "_"), sep="_"))) %>% 
        ungroup()
      
      long_species[[one_sample_type]] <- long_one 
      
      long_one <- long_species_sample_filtered %>% 
        group_by(division, class, order, family, genus) %>%
        summarize(min_pct_reads = min(n_reads, na.rm = TRUE)*100, 
                  mean_pct_reads = mean(n_reads, na.rm = TRUE)*100, 
                  max_pct_reads = max(n_reads, na.rm = TRUE)*100, 
                  n_samples = n()) %>% 
        rename_at(vars(matches("n_|pct_")), funs(str_c( .,str_replace_all(one_sample_type, " ", "_"), sep="_"))) %>% 
        ungroup()
      
      long_genus[[one_sample_type]] <- long_one 

      long_one <- long_species_sample_filtered %>%
        group_by(division, class) %>%
        summarize(min_pct_reads = min(n_reads, na.rm = TRUE)*100, 
                  mean_pct_reads = mean(n_reads, na.rm = TRUE)*100, 
                  max_pct_reads = max(n_reads, na.rm = TRUE)*100, 
                  n_samples = n())%>% 
        rename_at(vars(matches("n_|pct_")), funs( str_c(., str_replace_all(one_sample_type, " ", "_"), sep="_")))  %>% 
        ungroup()    
      
      long_class[[one_sample_type]] <- long_one 

  }

  long_class_comparison <- long_class %>% 
    purrr::reduce(full_join)
  
  long_genus_comparison <- long_genus %>% 
    purrr::reduce(full_join)

  long_species_comparison <- long_species %>% 
    purrr::reduce(full_join)

  # Class ----------
  
  cat("Class not found in one type of sample\n")
  cat("18S filter\n")
   kable(filter(long_class_comparison, is.na(n_samples_18S_filter)))
  cat("18S sort\n")
   kable(filter(long_class_comparison, is.na(n_samples_18S_sort)))
  cat("16S filter\n")
   kable(filter(long_class_comparison, is.na(n_samples_16S_plastid)))

  cat("Genus  found in the three types of samples\n")
  cat("18S filter\n")
   kable(filter(long_genus_comparison, (!is.na(n_samples_18S_filter)& 
                                        !is.na(n_samples_18S_sort) &
                                        !is.na(n_samples_16S_plastid) )))

  # Genus ----------
   
  cat("Genus only found in one type of sample\n")
  cat("18S filter\n")
   kable(filter(long_genus_comparison, (is.na(n_samples_18S_sort) &
                                        is.na(n_samples_16S_plastid) )))
  cat("18S sort\n")
   kable(filter(long_genus_comparison, (is.na(n_samples_18S_filter)&
                                        is.na(n_samples_16S_plastid) )))
  cat("16S filter\n")
   kable(filter(long_genus_comparison, (is.na(n_samples_18S_filter)& 
                                        is.na(n_samples_18S_sort)  )))
   

  cat("Species  found in the three types of samples\n")
  cat("18S filter\n")
   kable(filter(long_species_comparison, (!is.na(n_samples_18S_filter)& 
                                        !is.na(n_samples_18S_sort) &
                                        !is.na(n_samples_16S_plastid) )))
   
  long_genus_comparison_upset <- long_genus_comparison %>% 
    filter(!str_detect(genus, "_"))  %>% 
    select(genus, contains("samples")) %>%
    rename_at(vars(contains("n_")), funs(str_replace_all(.,c("n_samples_" = "", "_" = " ", "plastid" = "filter")))) 
   
  fig_upset_list[["method genus"]] <- plot_upset_long(long_genus_comparison_upset, "../fig/upset_method_genus.png")

  # Species ----------
  
   ggplot( long_species_comparison, aes(x=mean_pct_reads_18S_filter, y=mean_pct_reads_18S_sort, label=species, fill=class)) +
    geom_label(alpha= 0.5) +
    scale_x_log10() +
    scale_y_log10() +
    theme_bw() +
    labs(x = "18S sort", y="18S sort")  +
    scale_fill_manual(values = class_colors) +
    geom_abline(slope=1)
  
   ggplot( long_species_comparison, aes(x=mean_pct_reads_18S_filter, y=mean_pct_reads_16S_plastid, label=species, fill=class)) +
    geom_label(alpha= 0.5) +
    scale_x_log10() +
    scale_y_log10() +
    theme_bw() +
    labs(x = "18S sort", y="16S filter")  +
    scale_fill_manual(values = class_colors) +
    geom_abline(slope=1)
   

  cat("Species only found in one type of sample\n")
  cat("18S filter\n")
   kable(filter(long_species_comparison, (is.na(n_samples_18S_sort) &
                                        is.na(n_samples_16S_plastid) )))
  cat("18S sort\n")
   kable(filter(long_species_comparison, (is.na(n_samples_18S_filter)&
                                        is.na(n_samples_16S_plastid) )))
  cat("16S filter\n")
   kable(filter(long_species_comparison, (is.na(n_samples_18S_filter)& 
                                        is.na(n_samples_18S_sort)  )))
   
  long_species_comparison_upset <- long_species_comparison %>% 
    filter(!str_detect(genus, "_"))  %>% 
    select(species, contains("samples")) %>%
  rename_at(vars(contains("n_")), funs(str_replace(.,"n_samples_", ""))) 
  

  
   
  fig_upset_list[["method species"]] <- plot_upset_long(long_species_comparison_upset, "../fig/upset_method_species.png")



```

## Write comparison to Excel file

```{r, eval = FALSE}

   openxlsx::write.xlsx(list(long_class_comparison,long_genus_comparison,long_species_comparison ), 
                        file="../dada2/method_comparison.xlsx",
                        sheetName=c("Class","Genus", "Species"),
                        firstActiveRow = 2, colWidths ="auto", numFmt="0.00", zoom=75)

```


## Comparison of species found in the different fractions for 18S filter and 18S sorted   

* Compute table of number of samples for each species (rows) vs the three fractions (columns)

* Only keep
  * 2015 samples because it is the only dataset for which we have the three types of samples
  * Genera that do not contain _X


```{r  fig.height=12, fig.width=20}

long_species_fraction <- list()
long_class_fraction <- list()

  for (one_sample_type in sample_type[c(1,3)]) {
      long_one <- long_species_sample[[one_sample_type]] %>%
        filter(date >= as.Date("2015-01-01")) %>% 
        filter(!str_detect(genus, "_")) %>%  
        group_by(fraction_name_original, division, class, order, family, genus, species) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_species_fraction[[one_sample_type]] <- long_one 

      long_one <- long_class_sample[[one_sample_type]] %>%
        filter(date >= as.Date("2015-01-01")) %>% 
        group_by(fraction_name_original, division, class) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_class_fraction[[one_sample_type]] <- long_one 

  }

  long_class_fraction_comparison <- long_class_fraction %>% 
    purrr::reduce(full_join)
  long_species_fraction_comparison <- long_species_fraction %>% 
    purrr::reduce(full_join)
  
   kable(long_class_fraction_comparison)
   kable(long_species_fraction_comparison)
   
   fig_upset_list[["filter_sort species"]] <- plot_upset_long(long_species_fraction_comparison, "../fig/upset_filter_sort.png")

```


## Comparison of species found in the different fractions for 18S filter   
* Compute table of number of samples for each species (rows) vs the three fractions (columns)

* Only keep
  * Genera that do not contain _X
* We consider now ALL samples (will need to remove samples for which the > 20 um is missing)

```{r  fig.height=12, fig.width=20}

long_species_fraction <- list()
long_genus_fraction <- list()
long_class_fraction <- list()

  for (one_sample_type in sample_type[c(1)]) {
      long_one <- long_species_sample[[one_sample_type]] %>%
        filter(!str_detect(genus, "_")) %>%  
        group_by(fraction_name_original, division, class, order, family, genus, species) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_species_fraction[[one_sample_type]] <- long_one 
      
      long_one <- long_species_sample[[one_sample_type]] %>%
        filter(!str_detect(genus, "_")) %>%  
        group_by(fraction_name_original, division, class, order, family, genus) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_genus_fraction[[one_sample_type]] <- long_one 
      
      
      long_one <- long_class_sample[[one_sample_type]] %>%
        group_by(fraction_name_original, division, class) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_class_fraction[[one_sample_type]] <- long_one 

  }

  long_class_fraction_comparison <- long_class_fraction[[one_sample_type]]

  long_genus_fraction_comparison <- long_genus_fraction[[one_sample_type]]

  long_species_fraction_comparison <- long_species_fraction[[one_sample_type]] 

   kable(long_class_fraction_comparison)
   kable(long_genus_fraction_comparison)
   kable(long_species_fraction_comparison)

   fig_upset_list[["filter genus"]] <- plot_upset_long(long_genus_fraction_comparison, "../fig/upset_filter_genus.png")
   fig_upset_list[["filter species"]] <- plot_upset_long(long_species_fraction_comparison, "../fig/upset_filter_species.png")
```


## Comparison of species found in the different fractions for 18S sort   
* Compute table of number of samples for each species (rows) vs the three fractions (columns)

* Only keep
  * Genera that do not contain _X
* We consider now ALL samples (will need to remove samples for which the > 20 um is missing)

```{r  fig.height=12, fig.width=20}

long_species_fraction <- list()
long_class_fraction <- list()

  for (one_sample_type in sample_type[c(3)]) {
      long_one <- long_species_sample[[one_sample_type]] %>%
        filter(!str_detect(genus, "_")) %>%  
        group_by(fraction_name_original, division, class, order, family, genus, species) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_species_fraction[[one_sample_type]] <- long_one 

  long_one <- long_species_sample[[one_sample_type]] %>%
        filter(!str_detect(genus, "_")) %>%  
        group_by(fraction_name_original, division, class, order, family, genus) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_genus_fraction[[one_sample_type]] <- long_one 

      long_one <- long_class_sample[[one_sample_type]] %>%
        group_by(fraction_name_original, division, class) %>%
        summarize(n_samples = n()) %>% 
        pivot_wider(names_from = fraction_name_original, values_from = n_samples) %>% 
        ungroup()
      
      long_class_fraction[[one_sample_type]] <- long_one 

  }

  long_class_fraction_comparison <- long_class_fraction[[one_sample_type]]

  long_genus_fraction_comparison <- long_genus_fraction[[one_sample_type]]

  long_species_fraction_comparison <- long_species_fraction[[one_sample_type]] 

   kable(long_class_fraction_comparison)
   kable(long_genus_fraction_comparison)
   kable(long_species_fraction_comparison)
   
   fig_upset_list[["sort genus"]] <- plot_upset_long(long_genus_fraction_comparison, "../fig/upset_sort_genus.png")
   fig_upset_list[["sort species"]] <- plot_upset_long(long_species_fraction_comparison, "../fig/upset_sort_species.png")
```


# Figures

## Class legend

```{r}
  dummy <- long[["18S filter surface"]] %>% 
   count(class, wt= n_reads)

  legend_class <- ggplot(dummy, aes (x = class, y = n, fill = class)) +  
    geom_col() +
    scale_fill_manual(values = class_colors, drop = FALSE) +
    guides(fill = guide_legend(title.position="top",
                               ncol = 6, byrow = TRUE))
  
  legend_class
  
  legend_class <- ggpubr::get_legend(legend_class  + 
                      # create some space to the left of the legend    
                      theme(legend.box.margin = margin(t = 0, r = 0, b = 0, l = 0))
                    )
  
  ggpubr::as_ggplot(legend_class)
```


## List of genera

```{r, eval = FALSE}
  dummy <- long[["18S filter surface"]] %>%
   bind_rows(long[["18S sort surface"]]) %>% 
   bind_rows(long[["16S plastid surface"]]) %>% 
   count(class, genus, wt= n_reads) %>% 
   select(-n)

  # openxlsx::write.xlsx(dummy, "genus_colors.xlsx")
```


## Figure 2 - Treemap

### Cowplot - DO NOT USE

```{r fig.height=30, fig.width=15, eval = FALSE}

treemap_18S_filt <- cowplot::plot_grid(treemap_class[["18S filter 0.2 um surface"]]$gg + ggtitle("0.2 µm"),
                                       treemap_class[["18S filter 3 um surface"]]$gg + ggtitle("3 µm"),
                                       treemap_class[["18S filter 20 um surface"]]$gg + ggtitle("20 µm"), 
                                       ncol=1)

treemap_18S_sort <- cowplot::plot_grid(treemap_class[["18S sort pico surface"]]$gg,
                                       treemap_class[["18S sort nano surface"]]$gg,
                                       treemap_class[["fill in"]]$gg,
                                       ncol=1)

treemap_16S_filt <- cowplot::plot_grid(treemap_class[["16S plastid 0.2 um surface"]]$gg,
                                       treemap_class[["16S plastid 3 um surface"]]$gg,
                                       treemap_class[["16S plastid 20 um surface"]]$gg, 
                                       ncol=1)

fig_2 <- cowplot::plot_grid(legend_class, 
                            NULL,
                            bargraph_species[["18S filter surface"]]$gg + theme(legend.position="none"), 
                            treemap_18S_filt,
                            bargraph_species[["18S sort surface"]]$gg + theme(legend.position="none"), 
                            treemap_18S_sort,
                            bargraph_species[["16S plastid surface"]]$gg + theme(legend.position="none"), 
                            treemap_16S_filt,
                            ncol=2,rel_widths = c(3,1), rel_heights = c(1,3,3,3), align = "v")

fig_2
```

### Patchwork
```{r fig.height=30, fig.width=15}

treemap_18S_filt <- (treemap_class[["18S filter 0.2 um surface"]]$gg + ggtitle("0.2 µm")) /
                    (treemap_class[["18S filter 3 um surface"]]$gg + ggtitle("3 µm")) /
                    (treemap_class[["18S filter 20 um surface"]]$gg + ggtitle("20 µm"))

treemap_18S_sort <- (treemap_class[["18S sort pico surface"]]$gg + ggtitle("pico")) /
                    (treemap_class[["18S sort nano surface"]]$gg + ggtitle("nano")) /
                    plot_spacer()

treemap_16S_filt <- (treemap_class[["16S plastid 0.2 um surface"]]$gg + ggtitle("0.2 µm"))/
                    (treemap_class[["16S plastid 3 um surface"]]$gg + ggtitle("3 µm"))/
                    (treemap_class[["16S plastid 20 um surface"]]$gg + ggtitle("20 µm"))

fig_2 <- ggpubr::as_ggplot(legend_class) /
  
                            ( bargraph_species[["18S filter surface"]]$gg + theme(legend.position="none") + 
                            treemap_18S_filt + plot_layout(widths = c(1,1))) / 
                            (bargraph_species[["18S sort surface"]]$gg  + theme(legend.position="none") + 
                            treemap_18S_sort+ plot_layout(widths = c(1,1))) /
                            (bargraph_species[["16S plastid surface"]]$gg + theme(legend.position="none") + ggtitle("16S filter surface") + 
                            treemap_16S_filt+ plot_layout(widths = c(1,1))) +
                            plot_layout(heights = c(1,3,3,3))

fig_2
```

## Figure Supp - Treemap at genus level

```{r fig.height=15, fig.width=15}

treemap_18S_filt <- cowplot::plot_grid(treemap_genus[["18S filter 0.2 um surface"]]$gg + labs(title = "18S filter", subtitle =  "0.2 µm"),
                                       treemap_genus[["18S filter 3 um surface"]]$gg + labs(title = "", subtitle =  "3 µm") +
                                         theme(plot.title=element_blank()),
                                       treemap_genus[["18S filter 20 um surface"]]$gg + labs(title = "", subtitle =  "20 µm")+
                                         theme(plot.title=element_blank()), 
                                       ncol=1)

treemap_18S_sort <- cowplot::plot_grid(treemap_genus[["18S sort pico surface"]]$gg + labs(title = "18S sort", subtitle =  "pico"),
                                       treemap_genus[["18S sort nano surface"]]$gg + labs(title = "", subtitle =  "nano")+
                                         theme(plot.title=element_blank()),
                                       treemap_genus[["fill in"]]$gg,
                                       ncol=1)

treemap_16S_filt <- cowplot::plot_grid(treemap_genus[["16S plastid 0.2 um surface"]]$gg + labs(title = "16S filter", subtitle =  "0.2 µm"),
                                       treemap_genus[["16S plastid 3 um surface"]]$gg + labs(title = "", subtitle =  "3 µm")+
                                         theme(plot.title=element_blank()),
                                       treemap_genus[["16S plastid 20 um surface"]]$gg + labs(title = "", subtitle =  "20 µm")+
                                         theme(plot.title=element_blank()), 
                                       ncol=1)

fig_supp_treemap <- cowplot::plot_grid(treemap_18S_filt,
                            treemap_18S_sort,
                            treemap_16S_filt,
                           # labels = c("A" ,"B", "C"), label_x = 0.9,
                            ncol=3)

fig_supp_treemap
```


## Figure 3 - NMDS 18S filter

### Cowplot - DO NOT USE

```{r fig.height=12, fig.width=8, eval=FALSE}


fig_3 <- cowplot::plot_grid(nmds_sample[["18S filter surface"]]$gg_samples,
                            nmds_sample[["18S filter surface"]]$gg_taxa, 
                           labels = c("A" ,"B"), label_x = 0.9,
                            nrow=2,rel_widths = c(1,1))

fig_3
```

### Patchwork
```{r fig.height=12, fig.width=8}


fig_3 <- (nmds_sample[["18S filter surface"]]$gg_samples  + labs(title = "18S filter", shape="fraction" )) /
                            (nmds_sample[["18S filter surface"]]$gg_taxa + ggtitle(""))

fig_3
```

## Figure sup - NMDS 16S plastid

### Cowplot - DO NOT USE

```{r fig.height=12, fig.width=8, eval = FALSE}


fig_NMDS_plastid <- cowplot::plot_grid(nmds_sample[["16S plastid surface"]]$gg_samples + ggtitle("16S filter"),
                            nmds_sample[["16S plastid surface"]]$gg_taxa + ggtitle(""), 
                           labels = c("A" ,"B"), label_x = 0.9,
                            nrow=2,rel_widths = c(1,1))

fig_NMDS_plastid
```

### Patchwork
```{r fig.height=12, fig.width=8}


fig_NMDS_plastid <- (nmds_sample[["16S plastid surface"]]$gg_samples + labs(title = "16S filter", shape="fraction" ) ) /
                    nmds_sample[["16S plastid surface"]]$gg_taxa + ggtitle("") 

fig_NMDS_plastid
```

## Figure 4 - Heatmap 18S filter

### Cowplot
```{r fig.height=12, fig.width=25, eval = FALSE}
fig_4 <- cowplot::plot_grid(heatmap_class_selected[["18S filter 0.2 um surface"]],
                           heatmap_class_selected[["18S filter 3 um surface"]],
                           heatmap_class_selected[["18S filter 20 um surface"]],
                           heatmap_species_selected[["18S filter 0.2 um surface"]],
                           heatmap_species_selected[["18S filter 3 um surface"]],
                           heatmap_species_selected[["18S filter 20 um surface"]], 
                           labels = c("A" ,"B","C", "D","E", "F"), label_x = 0.9,
                           nrow=2, ncol=3)
fig_4
```

### Patchwork
```{r fig.height=12, fig.width=25}
fig_4 <- heatmap_class_selected[["18S filter 0.2 um surface"]]+ 
  labs(title ="0.2 µm") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank())  +
  
  heatmap_class_selected[["18S filter 3 um surface"]]+ 
  labs(title ="3 µm") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank())  +
  
  heatmap_class_selected[["18S filter 20 um surface"]]+ 
  labs(title ="20 µm")+ 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank()) +
  
  heatmap_species_selected[["18S filter 0.2 um surface"]]+ 
  labs(title ="")+ 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank()) +
  
  heatmap_species_selected[["18S filter 3 um surface"]] +
  labs(title ="")+ 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank()) +
  
  heatmap_species_selected[["18S filter 20 um surface"]]+ 
  labs(title ="") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank())  +
  
  plot_layout(nrow = 2, byrow = TRUE, guides = "collect") 

fig_4
```
## Figure sup - Heatmap 16S filter


### Patchwork
```{r fig.height=12, fig.width=25}
fig_heatmap_16S <- heatmap_class_selected[["16S plastid 0.2 um surface"]]+ 
  labs(title ="0.2 µm") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank())  +
  
  heatmap_class_selected[["16S plastid 3 um surface"]]+ 
  labs(title ="3 µm") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank())  +
  
  heatmap_class_selected[["16S plastid 20 um surface"]]+ 
  labs(title ="20 µm")+ 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank()) +
  
  plot_layout(nrow = 1, byrow = TRUE, guides = "collect") 

fig_heatmap_16S
```

## Figure sup - Heatmap 18S sort
### Cowplot - DO NOT USE
```{r fig.height=10, fig.width=15, eval = FALSE}
fig_heatmap_sort <- cowplot::plot_grid(heatmap_class_selected[["18S sort pico surface"]],
                           heatmap_class_selected[["18S sort nano surface"]],
                           heatmap_species_selected[["18S sort pico surface"]],
                           heatmap_species_selected[["18S sort nano surface"]], 
                           labels = c("A" ,"B","C", "D"), label_x = 0.9,
                           nrow=2, ncol=2)
fig_heatmap_sort
```
### Patchwork

```{r fig.height=10, fig.width=15}
fig_heatmap_sort <- heatmap_class_selected[["18S sort pico surface"]]+ 
  labs(title ="pico") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank()) +
  
  heatmap_class_selected[["18S sort nano surface"]]+ 
  labs(title ="nano") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank()) + 
  
  heatmap_species_selected[["18S sort pico surface"]]+ 
  labs(title ="") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank()) + 
  
  heatmap_species_selected[["18S sort nano surface"]] + 
  labs(title ="") + 
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank())   +
  
  plot_layout(nrow = 2, byrow = TRUE, guides = "collect") 

fig_heatmap_sort
```


## Figure 6 - Vertical Profile

### Cowplot - DO NOT USE
```{r fig.height=12, fig.width=6, eval = FALSE}

legend <- cowplot::get_legend( bargraph_sample[["18S filter 0.2 um profile"]] + 
                      # create some space to the left of the legend    
                      theme(legend.box.margin = margin(0, 0, 0, 20))
                    )
fig_6 <- cowplot::plot_grid(bargraph_sample[["18S filter 0.2 um profile"]]+ theme(legend.position="none"),legend,
                            bargraph_sample[["18S filter 3 um profile"]] + theme(legend.position="none"),NULL,
                            bargraph_sample[["18S filter 20 um profile"]]+ theme(legend.position="none"),NULL,
                            labels = c("" ,"A","", "B","", "C"), label_x = 0.9,
                            nrow=3, ncol=2, rel_widths = c(3, 1.7))
fig_6
```

### Patchwork

```{r fig.height=8, fig.width=10}

                  
fig_6 <-  ggpubr::as_ggplot(legend_class) / (
  bargraph_sample[["18S filter 0.2 um profile"]]+ theme(legend.position="none") + 
  labs(title ="0.2 µm") +  
  bargraph_sample[["18S filter 3 um profile"]]+ theme(legend.position="none", axis.title.y=element_blank()) + 
  labs(title ="3 µm") + 
  bargraph_sample[["18S filter 20 um profile"]]+ theme(legend.position="none", axis.title.y=element_blank())+ 
  labs(title ="20 µm")  )  +
  plot_layout(heights = c(0.3,1))
fig_6
```


## Figure X - Upset R

* Problem is that upset does not provide a ggplot object.  So need to save as png
* Use ggplotify library : https://cran.r-project.org/web/packages/ggplotify/vignettes/ggplotify.html
* But it heats a bit the plots

```{r fig.height=10, fig.width=6}

p1 <- cowplot::ggdraw() + cowplot::draw_image("../fig/upset_method_genus.png")
p2 <- cowplot::ggdraw() + cowplot::draw_image("../fig/upset_filter_genus.png")
p3 <- cowplot::ggdraw() + cowplot::draw_image("../fig/upset_sort_genus.png")

fig_X_upset <- cowplot::plot_grid(p1, p2, p3, labels = c("A", "B", "C"),label_x = 0.9,
                                  nrow=3, ncol=1)
fig_X_upset
```

### Patchword - DO NOT USE
```{r fig.height=10, fig.width=10, eval = FALSE}
  ggplotify::as.ggplot(fig_upset_list[["method"]]) / 
  ggplotify::as.ggplot(fig_upset_list[["filter"]]) /  
  ggplotify::as.ggplot(fig_upset_list[["sort"]])
```

## Save figures
```{r, eval = FALSE}

fig_path <- function (file) str_c("C:/daniel.vaulot@gmail.com/Papers/2020 Trefault Antarctic/paper_antarctic_overleaf/fig/", file)


ggsave(plot= fig_2 , filename=fig_path("Figure_treeplot.pdf"), 
       width = 16 , height = 25, scale=1.8, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_3 , filename=fig_path("Figure_NMDS.pdf"), 
       width = 12 , height = 18, scale=1.75, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_4 , filename=fig_path("Figure_heatmap.pdf"), 
       width = 20 , height = 10, scale=2.2, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_6 , filename=fig_path("Figure_vertical_profile.pdf"), 
       width = 13 , height = 8, scale=2.2, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_X_upset , filename=fig_path("Figure_upset.pdf"), 
       width = 8 , height = 17, scale=2.2, units="cm", useDingbats=FALSE) 

ggsave(plot= fig_NMDS_plastid , filename=fig_path("Figure-sup_NMDS_16S_plastid.pdf"), 
       width = 12 , height = 18, scale=1.75, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_heatmap_sort , filename=fig_path("Figure-sup_heatmap_18S_sort.pdf"), 
       width = 12 , height = 10, scale=2, units="cm", useDingbats=FALSE) 
ggsave(plot= fig_heatmap_16S , filename=fig_path("Figure-sup_heatmap_16S.pdf"), 
       width = 12 , height = 5, scale=3, units="cm", useDingbats=FALSE)
ggsave(plot= fig_supp_treemap , filename=fig_path("Figure-sup_treemap_genus.pdf"), 
       width = 12 , height = 12, scale=2, units="cm", useDingbats=FALSE)



```

# Saving Tables to Latex

```{r}

library(xtable)
```

## Define function to format the tables
* Not used 
    * "X" = "\\\\cellcolor{gray}"
```{r, eval=FALSE}
sanitize.italics <- function(str) {
  str_replace_all(str, c("_" = " ", 
                         # "ital\\{" = "\\\\textit{", 
                         "Â°" = "\\\\degree",
                         "±"="$\\pm$"))
}
```

## File path

```{r, eval=FALSE}
path_table <- function(file_table) {str_c("C:/daniel.vaulot@gmail.com/Papers/2020 Trefault Antarctic/paper_antarctic_overleaf/tables/", file_table)}

file_main <- path_table("Tables.xlsx")
```

## Read legends
```{r, eval=FALSE}
legends <- readxl::read_excel(path=file_main, sheet="Table legends")

```


## Table - Data sets

### Using Xtable
```{r, eval=FALSE}
one_sheet <- "Table data sets"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A1:G4")
table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","c","c", "c","c","c", "c", "c"),
                        digits=0)
print(table,  scalebox = 0.85, 
      caption.placement = "top",
      include.rownames = FALSE, 
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)
```

### Using KableXtra
```{r}
one_sheet <- "Table data sets"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A1:G4")

table_latex <- kbl(table, 
                   caption = one_caption, 
                   label=str_remove(one_label,"tab:"), 
                   booktabs = T, 
                   format = "latex", 
                   escape = F) %>% 
  kable_styling(latex_options = c("scale_down"),
                position="center")

cat(table_latex, file = path_table(one_file))
```


## Table - Samples

### Using Xtable
```{r, eval=FALSE}
one_sheet <- "Table samples"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

addtorow <- list()
addtorow$pos <- list(0, 0)
addtorow$command <- c( "Date & Season & CTD & Chl & FCM & Nutrients & Profile & \\multicolumn{3}{c}{18S rRNA filter} & \\multicolumn{3}{c}{16S rRNA filter}  & \\multicolumn{2}{c}{18S rRNA sort} \\\\\n", 
"& & & & & & & 0.2 $\\mu$m  & 3 $\\mu$m  & 20 $\\mu$m  & 0.2 $\\mu$m  & 3 $\\mu$m  & 20 $\\mu$m  & Pico  & Nano \\\\\n")


table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A3:O22", col_names = FALSE)
table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","l","c", "c","c","c", "c", "c", "c","c","c", "c","c","c", "c", "c"),
                        digits=0)
print(table,  scalebox = 0.75, 
      caption.placement = "top",
      add.to.row = addtorow, 
      include.colnames = FALSE,
      include.rownames = FALSE, 
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)
```

### Using KableXtra
```{r}
one_sheet <- "Table samples"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]
one_colname <- c("Date", "Season", "CTD", "Chl", "FCM", "Nutrients", "Profile", 
                 "0.2 $\\mu$m", "3 $\\mu$m", "20 $\\mu$m", 
                 "0.2 $\\mu$m", "3 $\\mu$m", "20 $\\mu$m", 
                 "Pico ", "Nano")

table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A3:O22", col_names = FALSE)

table_latex <- kbl(table, 
                   caption = one_caption, 
                   label=str_remove(one_label,"tab:"), 
                   col.names = one_colname,
                   align = c("l", rep('c',times=14)),
                   booktabs = T, 
                   format = "latex", 
                   escape = F) %>% 
  kable_styling(latex_options = c("scale_down"),
                position="center") %>% 
  add_header_above(c(" "= 7 , 
                     "18S rRNA filter" = 3, 
                     "16S rRNA filter" = 3,
                     "18S rRNA sort" = 2)) 

cat(table_latex, file = path_table(one_file))
```


## Table - Vertical Profile
```{r, eval=FALSE}
one_sheet <- "Table vertical profile"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A1:K6")
table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","c","c", "c","c","c", "c", "c", "c","c","c", "c"),
                        digits=c(0,0,2,1,2,1,2,2,1,0,0,0))
print(table,  scalebox = 1, 
      caption.placement = "top",
      include.rownames = FALSE, 
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)
```

## Table - ANOSIM
```{r, eval=FALSE}
one_sheet <- "Table ANOSIM"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A1:D5")
table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","c","c", "c","c"),
                        digits=3)
print(table,  scalebox = 1, 
      caption.placement = "top",
      include.rownames = FALSE, 
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)
```

## Table - Species surface
```{r, eval=FALSE}
one_sheet <- "Table species surface"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

addtorow <- list()
addtorow$pos <- list(0, 0, 0, 0, 0, 0, 0)
addtorow$command <- c( 
  # Header for first page
  "Division & Class & Species & \\multicolumn{4}{c}{18S rRNA filter} & \\multicolumn{4}{c}{16S rRNA plastid filter}  & \\multicolumn{4}{c}{18S rRNA sort} \\\\\n", 
  "& & & min & mean & max & n & min & mean & max & n & min & mean & max & n \\\\\n",
  "\\endfirsthead \n \\hline \n",
  
  # Header for other pages
  "Division & Class & Species & \\multicolumn{4}{c}{18S filter} & \\multicolumn{4}{c}{16S filter}  & \\multicolumn{4}{c}{18S sort} \\\\\n", 
  "& & & min & mean & max & n & min & mean & max & n & min & mean & max & n \\\\\n",
  "\\hline \n \\endhead \n",
  
  # Footers
  "\\hline \n \\endfoot \n  \\endlastfoot \n"
)



table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A3:O129", col_names = FALSE) 


table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","l","l", "l","c","c", "c", "c", "c","c","c", "c","c","c", "c", "c"),
                        digits= c(0, 0, 0,0, 2,2,2,0,2,2,2,0,2,2,2,0 ))
print(table,  
      # scalebox = 0.75, 
      caption.placement = "top",
      add.to.row = addtorow, 
      include.colnames = FALSE,
      include.rownames = FALSE, 
      tabular.environment = "longtable",
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)
```


## Table - Species euphotic
```{r, eval=FALSE}
one_sheet <- "Table species euphotic"
one_legend <- filter(legends, table == one_sheet)
one_caption <- one_legend$legend[1]
one_label <- one_legend$label[1]
one_file <- one_legend$file_name[1]

addtorow <- list()
addtorow$pos <- list(0, 0, 0, 0, 0)
addtorow$command <- c( 
  # Header for first page
  "Division & Class & Species \\\\\n", 
  "\\endfirsthead \n \\hline \n",
  
  # Header for other pages
  "Division & Class & Species \\\\\n", 
  "\\hline \n \\endhead \n",
  
  # Footers
  "\\hline \n \\endfoot \n  \\endlastfoot \n"
)



table <- readxl::read_excel(path=file_main, sheet=one_sheet, range= "A2:C44", col_names = FALSE) 


table <- xtable::xtable(table, label=one_label, 
                        caption=one_caption, 
                        align = c("l","l","l", "l")
                        )
print(table,  
      # scalebox = 0.75, 
      caption.placement = "top",
      add.to.row = addtorow, 
      include.colnames = FALSE,
      include.rownames = FALSE, 
      tabular.environment = "longtable",
      file=path_table(one_file),
      sanitize.text.function = sanitize.italics)

```

